File: PinpinboxShareExtension/ShareItem.m
//
//  ShareItem.m
//  PinpinboxShareExtension
//
//  Created by Antelis on 2018/12/17.
//  Copyright © 2018 Angus. All rights reserved.
//

#import "ShareItem.h"
#import "NSURL+Param.h"
#import <MobileCoreServices/MobileCoreServices.h>
#import <AVFoundation/AVFoundation.h>

@interface ShareItem()
@property (nonatomic) id<ItemContentDelegate> itemDelegate;
@end

@implementation ShareItem
- (id)initWithItemProvider:(NSItemProvider *)item type:(NSString *)type itemDelegate:(id<ItemContentDelegate>)itemDelegate{
    self = [super init];
    if (self) {
        self.itemDelegate = itemDelegate;
        self.hasVideo = NO;
        self.thumbIsDark = NO;
        self.vidDuration = 0;
        self.objType = type;
        self.shareItem = item;
        
        [self postLoadShareItem];
    }
    return self;
}
- (void)postLoadShareItem {
    __block typeof(self) wself = self;
    [self.shareItem loadItemForTypeIdentifier:self.objType options:nil completionHandler:^(id<NSSecureCoding>  _Nullable item, NSError * _Null_unspecified error) {
        if (!error) {
            if ([wself.objType isEqualToString:(__bridge  NSString *)kUTTypeText]) {

                @try {
                    
                    NSString *text = (NSString *)item;
                    if (![text hasPrefix:@"file://"]) {
                        NSURL *url = [NSURL URLWithString:text];
                        [wself setUrl:url];
                    }
                    
                } @catch (NSException *exception) {
                    NSLog(@"Failed to retrieve URL %@",[exception description]);
                }
            } else {
                NSURL *url = (NSURL *)item;
                if (url && [url isKindOfClass:[NSURL class]])
                    [wself setUrl:url];
            }
                
            
            [wself tryLoadThumbnail];
        } else {
            
        }
        
    }];
}
- (void)tryLoadThumbnail {
    
    __block typeof(self) wself = self;
    if ([_objType isEqualToString:(__bridge NSString *)kUTTypeURL] ||
        [_objType isEqualToString:(__bridge NSString *)kUTTypeText]) {
        NSString *videoID = nil;
        NSString *path = [_url absoluteString];
        if ([path containsString:@"youtu"]) {
            
            if ([_url.host containsString:@"youtube.com"]) {
                videoID = [_url queryParam:@"v"];
            } else if ([_url.host containsString:@"youtu.be"]) {
                videoID = [_url lastPathComponent];
            }
            NSString *th = [NSString stringWithFormat:@"http://img.youtube.com/vi/%@/hqdefault.jpg",videoID];
            self.thumbURL = [NSURL URLWithString:th];
            self.hasVideo = YES;
        } else if ([_url.absoluteString containsString:@"vimeo"]) {
            self.hasVideo = YES;
            NSString *videoPath = _url.lastPathComponent;
            NSString *realLink = [NSString stringWithFormat:@"https://vimeo.com/api/oembed.json?url=https://vimeo.com/%@&width=960",videoPath];
            NSURL *url = [NSURL URLWithString:[realLink stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
            NSData *data = [NSData dataWithContentsOfURL:url];
            if (url  && data) {
                NSError *e1 = nil;
                NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:0 error:&e1];
                if (!e1 && dict ) {
                    NSString *tpath = dict[@"thumbnail_url"];
                    if (tpath) {
                        self.thumbURL = [NSURL URLWithString:tpath];
                        return;
                    }
                    
                }
                
            }
            //  nothing found
            self.thumbnail = [UIImage imageNamed:@"videobase.jpg"];
            
        } else if ([_url.absoluteString containsString:@"facebook"]) {
            self.thumbnail = [UIImage imageNamed:@"videobase.jpg"];
            self.hasVideo = YES;
        } else {
            self.hasVideo = NO;
            if ([self.shareItem hasItemConformingToTypeIdentifier:@"public.file-url"] && [self.shareItem hasItemConformingToTypeIdentifier:@"public.image"]) {
                self.objType = (__bridge NSString *)kUTTypeImage;
                [self setThumbnail:[UIImage imageWithContentsOfFile:[self.url path]]];
                
            } else {
                //  other invalid format : ordinary URL, pure text, or text file
               // [self loadThumbnailOtherway];
                if (self.itemDelegate)
                    [self.itemDelegate processInvalidItem:self];
            }
        }
    } else if ([_objType isEqualToString:(__bridge NSString *)kUTTypeMovie]){
        self.hasVideo = YES;//[_objType isEqualToString:(__bridge NSString *)kUTTypeMovie];
        [self.shareItem loadPreviewImageWithOptions:nil completionHandler:^(id<NSSecureCoding>  _Nullable item, NSError * _Null_unspecified error) {
            if (!error && item) {
                [wself setThumbnail:(UIImage *)item];
            } else {
                //[wself setThumbnail:[UIImage imageNamed:@"videobase.jpg"]];
            }
        }];
    } else if ([_objType isEqualToString:(__bridge NSString *)kUTTypeImage]) {
        self.hasVideo = NO;
        self.thumbnail = [UIImage imageWithContentsOfFile:[self.url path]];
    } else if ([_objType isEqualToString:(__bridge NSString *)kUTTypePDF]) {
        self.hasVideo = NO;
        [self loadThumbnailOtherway];
    }
}
- (void)loadThumbnailOtherway {
    __block typeof(self) wself = self;
    [self.shareItem loadPreviewImageWithOptions:nil completionHandler:^(id<NSSecureCoding>  _Nullable item, NSError * _Null_unspecified error) {
        if (!error && item) {
            [wself setThumbnail:(UIImage *)item];
        } else {
            
            [wself.url startAccessingSecurityScopedResource];
            
            NSFileCoordinator *coordinator = [[NSFileCoordinator alloc] init];
            __block NSError *error;
            [coordinator coordinateReadingItemAtURL:self.url options:0 error:&error byAccessor:^(NSURL *newURL) {
                NSDictionary *thumbs = nil;
                thumbs = [newURL resourceValuesForKeys:@[NSURLThumbnailDictionaryKey] error:&error]; //getResourceValue:&thumbs forKey:NSURLThumbnailDictionaryKey error:&error];
                if (thumbs && thumbs[NSURLThumbnailDictionaryKey]) {
                    NSLog(@"%@",thumbs);
                    NSDictionary *t =  thumbs[NSURLThumbnailDictionaryKey];
                    UIImage *tt = t[NSThumbnail1024x1024SizeKey];
                    if (tt)
                        [wself setThumbnail:tt];
                    else
                        [wself setThumbnail:[UIImage imageNamed:@"videobase.jpg"]];
                } else
                    [wself setThumbnail:[UIImage imageNamed:@"videobase.jpg"]];
            }];
            
            [wself.url stopAccessingSecurityScopedResource];
            
            
        }
    }];
}
- (void)setUrl:(NSURL *)url {
    _url = url;
    
    if ([_objType isEqualToString:(__bridge NSString *)kUTTypeMovie]) {
        AVURLAsset *sourceAsset = [AVURLAsset URLAssetWithURL:url options:nil];
        CMTime duration = sourceAsset.duration;
        _vidDuration = CMTimeGetSeconds(duration);
        if (_vidDuration >= 31.0) {
            if (self.itemDelegate)
                [self.itemDelegate processInvalidItem:self];
            return;
        }
        AVAssetImageGenerator *generator = [AVAssetImageGenerator assetImageGeneratorWithAsset:sourceAsset];
        NSError *err = nil;
        
        CGImageRef c = [generator copyCGImageAtTime:CMTimeMake(1.0, 1000) actualTime:nil error:&err];
        if (!err && c) {
            self.thumbnail = [UIImage imageWithCGImage:c];
            [self inspectThumbnailTone:self.thumbnail];
        }
        
    }
}
- (void)setThumbnail:(UIImage *)thumbnail  {
    _thumbnail = thumbnail;
    [self inspectThumbnailTone:thumbnail];
}
- (void)inspectThumbnailTone:(UIImage *)thumbnail {
    if (!thumbnail) return;
    __block typeof(self) wself = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        CGDataProviderRef provider = CGImageGetDataProvider(thumbnail.CGImage);
        CFDataRef data = CGDataProviderCopyData(provider);
        if (data) {
            UInt8 *ptr = (UInt8 *)CFDataGetBytePtr(data);
            NSInteger length = CFDataGetLength(data);
            int th = (thumbnail.size.height*thumbnail.size.width*45) /100;
            int darkp = 0;
            double luma = 0.0;
            for (int i = 0 ; i < length; i+=4) {
                UInt8 r = ptr[i];
                UInt8 g = ptr[i+1];
                UInt8 b = ptr[i+2];
                luma = (0.2126 * (double)r+0.7152 *(double)g+0.0722 * (double)b);
                if (luma < 150) {
                    darkp += 1;
                    if (darkp > th) {
                        
                        CFRelease(data);
                        wself.thumbIsDark = YES;
                        
                        return;
                    }
                    
                }
            }
            
        }
        
        CFRelease(data);
        wself.thumbIsDark = NO;
    });
}
- (void)loadThumbnailWithPostload:(id<ItemPostLoadDelegate>) postload {
    __block typeof(self) wself = self;
    if (_thumbnail) {
        dispatch_async(dispatch_get_main_queue(), ^{
            //imgView.image = wself.thumbnail;
            if (postload)
                [postload loadCompleted:wself.thumbnail type:wself.objType hasVideo:wself.hasVideo isDark:wself.thumbIsDark];
        });
        
    } else if (_thumbURL) {
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSData *data = [NSData dataWithContentsOfURL:wself.thumbURL];
            if (data) {
                UIImage *image = [UIImage imageWithData:data];
                dispatch_async(dispatch_get_main_queue(), ^{
                    wself.thumbnail = image;
                    if (postload)
                        [postload loadCompleted:wself.thumbnail type:wself.objType hasVideo:self.hasVideo isDark:self.thumbIsDark];
                });
            }
        });
    } else {
        [self postLoadShareItem];
        [self performSelector:@selector(loadThumbnailWithPostload:) withObject:postload afterDelay:3];
    }
}
@end

// 1. Vimeo has redirect URL in app ?? found no other cases
// XXX 2. https://vimeo.com/api/oembed.json?url=[URL]&width=960 seemed returned JSON!
// 3. YT live test is necessary
// 4. PDF upload process
// slide net ??? https://www.slideshare.net/developers
//  new album

File: PinpinboxShareExtension/ShareItem.swift
//
//  ShareItem.swift
//  PinpinboxShareExtension
//
//  Created by Antelis on 2019/3/19.
//  Copyright © 2019 Angus. All rights reserved.
//

import Foundation
import UIKit
import MobileCoreServices
import AVFoundation

extension URL {
    
    func queryParam(_ param : String) -> String {
        if let components = URLComponents(url: self, resolvingAgainstBaseURL: false), let items = components.queryItems {
            let result = items.filter { (item) -> Bool in
                return item.name == param
            }
            if let item = result.first, let value = item.value {
                return value
            }
        }
        return ""
    }
    
    func containsString(_ str: String) -> Bool {
        return self.path.contains(str)
        
    }
}

class ShareItem : NSObject {
    open var url : URL? {
        didSet {
            if let url = self.url, let objtype = self.objType, objtype == kUTTypeMovie as String {
                let sourceasset = AVURLAsset(url: url)
                let duration = sourceasset.duration
                self.vidDuration = Int(duration.seconds)
                
                if let d = self.itemDelegate, vidDuration > 31 {
                    d.processInvalidItem(self)
                    return;
                }
                let generator = AVAssetImageGenerator(asset: sourceasset)
                do {
                    let c = try generator.copyCGImage(at: CMTimeMake(value: 1, timescale: 1000), actualTime: nil)
                    self.thumbnail = UIImage(cgImage: c)
                    guard let img = self.thumbnail else {return}
                    self.inspectThumbnailTone(img)
                } catch let error {
                    print("Error in load asset image : \(error.localizedDescription)")
                }
            }

        }
        
    }
    var thumbURL : URL?
    var thumbnail: UIImage? {
        didSet {
            guard let img = self.thumbnail else {return}
            self.inspectThumbnailTone(img)
        }
    }
    var objType : String?
    var shareItem : NSItemProvider?
    var hasVideo : Bool = false
    var thumbIsDark : Bool = false
    var vidDuration: Int = 0
    var taskId: String?
    
    private var itemDelegate: ItemContentDelegate?
    
    convenience init(_ item: NSItemProvider?,_ type: String,_ delegate: ItemContentDelegate? ) {
        self.init()
        self.shareItem = item
        self.itemDelegate = delegate
    
        self.hasVideo = false
        self.thumbIsDark = false
        self.vidDuration = 0
        self.objType = type
        self.shareItem = item
        self.taskId = UUID.init().uuidString
        
        postLoadShareItem()
    }
    
    private func inspectThumbnailTone(_ image : UIImage) {
        
    }
    
    private func loadThumbnailWithPostload(_ postload: ItemPostLoadDelegate?) {
        guard let load = postload,  let otype = self.objType else {return }
        if let thumbnail = thumbnail {
            DispatchQueue.main.async {
                load.loadCompletedWith(thumbnail, otype, self.hasVideo, isDark: self.thumbIsDark)
            }
        } else if let thumbURL = self.thumbURL {
            DispatchQueue.global(qos: .default).async {
                do {
                    let data = try Data(contentsOf: thumbURL)
                    DispatchQueue.main.async {
                        if let img = UIImage(data: data) {
                            self.thumbnail = img
                            load.loadCompletedWith(img, otype, self.hasVideo, isDark: self.thumbIsDark)
                        }
                    }
                    
                } catch {
                    
                }
            }
        } else {
            self.postLoadShareItem()
            //[self performSelector:@selector(loadThumbnailWithPostload:) withObject:postload afterDelay:3];
        }
        
    }
    
    private func tryLoadThumbnail() {
        if let objtype = self.objType, let url = self.url {
            if objtype == kUTTypeURL as String || objtype == kUTTypeText as String {
                var videoID : String?
                if url.containsString("youtu") {
                    if let host = url.host, host.contains("youtube.com") {
                        videoID = url .queryParam("v")
                    } else if let host = url.host, host.contains("youtu.be") {
                        videoID = url.lastPathComponent
                    }
                    if let videoID = videoID {
                        let th = String(format:"http://img.youtube.com/vi/%@/hqdefault.jpg" , videoID)
                        self.thumbURL = URL(string: th)
                    } else {
                        self.thumbnail = UIImage(named:"videobase.jpg")
                    }
                    self.hasVideo = true
                    
                } else if url.containsString("vimeo") {
                    self.hasVideo = true
                    let videoPath = url.lastPathComponent
                    let realLink = String(format:"https://vimeo.com/api/oembed.json?url=https://vimeo.com/%@&width=960", videoPath)
                    if let encoded = realLink.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed), let vurl = URL(string: encoded) {
                        do {
                            let data = try Data(contentsOf: vurl)
                        
                            let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
                            
                            if let json = json, let p = json["thumbnail_url"], let p1 = p as? String {
                                self.thumbURL = URL(string: p1)
                                return
                            }
                            
                        } catch {
                            
                        }
                    }
                    
                    //  nothing found
                    self.thumbnail = UIImage(named:"videobase.jpg")
                    
                } else if url.containsString("facebook") {
                    self.thumbnail = UIImage(named:"videobase.jpg")
                    self.hasVideo = true
                } else {
                    self.hasVideo = false
                    if let item = self.shareItem {
                        if item.hasItemConformingToTypeIdentifier("public.file-url") && item.hasItemConformingToTypeIdentifier("public.image") {
                            self.objType = kUTTypeImage as String
                            self.thumbnail = UIImage(contentsOfFile: url.path)
                        } else if let delegate = self.itemDelegate{
                            delegate.processInvalidItem(self)
                        }
                    }
                }
            } else if objtype == kUTTypeMovie as String, let item = self.shareItem {
                self.hasVideo = true
                item.loadPreviewImage(options: [:]) { (coding, error) in
                    if let img = coding as? UIImage {
                        self.thumbnail = img
                    }
                }
                
            } else if objtype == kUTTypeImage as String {
                self.hasVideo = false
                self.thumbnail = UIImage(contentsOfFile: url.path)
            } else if objtype == kUTTypePDF as String {
                self.hasVideo = false;
                loadThumbnailOtherway()
            }
            
        }
    }
    
    private func loadThumbnailOtherway() {
        if let item = shareItem {
            item.loadPreviewImage(options: [:]) { (coding, error) in
                if let _ = coding, let image = coding as? UIImage {
                    self.thumbnail = image
                } else if let url = self.url {
                    if url.startAccessingSecurityScopedResource() {
                        let cr = NSFileCoordinator()
                        cr.coordinate(readingItemAt: url, options: [.immediatelyAvailableMetadataOnly], error: nil, byAccessor: { (newurl) in
                            if let res = try? newurl.resourceValues(forKeys: [.thumbnailDictionaryKey]),
                                let dict = res.thumbnailDictionary {
                                
                                let image = dict[.NSThumbnail1024x1024SizeKey]
                                self.thumbnail = image
                                return
                            }
                            self.thumbnail = UIImage(named:"videobase.jpg")
                        })
                    }
                    url.stopAccessingSecurityScopedResource()
                }
            }
        }
    }
    
    private func postLoadShareItem() {
        guard let item = self.shareItem, let objtype = self.objType else { return }
        item.loadItem(forTypeIdentifier:objtype , options: nil) { (coding, error) in
            if error == nil {
                if objtype == kUTTypeText as String {
                    if let text = coding as? String {
                        if !text.hasPrefix("file://") {
                            if let u = URL(string: text) {
                                self.url = u
                            }
                        }
                    } else if let url = coding as? URL {
                        self.url = url
                    }
                    
                    
                }
                self.tryLoadThumbnail()
            }
            
        }
    }
        
}
